scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,316 @@
+import heapq
+import itertools
+
+from dataclasses import dataclass
+from scipy._lib._util import MapWrapper
+
+import numpy as np
+
+from scipy.integrate._rules import (
+    ProductNestedFixed,
+    GaussKronrodQuadrature,
+    GenzMalikCubature,
+)
+from scipy.integrate._rules._base import _subregion_coordinates
+
+
+__all__ = ['cubature']
+
+
+@dataclass
+class CubatureRegion:
+    estimate: np.ndarray
+    error: np.ndarray
+    a: np.ndarray
+    b: np.ndarray
+
+    def __lt__(self, other):
+        # Consider regions with higher error estimates as being "less than" regions with
+        # lower order estimates, so that regions with high error estimates are placed at
+        # the top of the heap.
+
+        return np.max(np.abs(self.error)) > np.max(np.abs(other.error))
+
+
+@dataclass
+class CubatureResult:
+    estimate: np.ndarray
+    error: np.ndarray
+    status: str
+    regions: list[CubatureRegion]
+    subdivisions: int
+    atol: float
+    rtol: float
+
+
+def cubature(f, a, b, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
+             args=(), workers=1):
+    r"""
+    Adaptive cubature of multidimensional array-valued function.
+
+    Given an arbitrary integration rule, this function returns an estimate of the
+    integral to the requested tolerance over the region defined by the arrays `a` and
+    `b` specifying the corners of a hypercube.
+
+    Convergence is not guaranteed for all integrals.
+
+    Parameters
+    ----------
+    f : callable
+        Function to integrate. `f` must have the signature::
+            f(x : ndarray, \*args) -> ndarray
+
+        `f` should accept arrays ``x`` of shape::
+            (npoints, ndim)
+
+        and output arrays of shape::
+            (npoints, output_dim_1, ..., output_dim_n)
```

Comment:
I suspect `::` needs a blank line to render properly, here and elsewhere:

```suggestion

            (npoints, output_dim_1, ..., output_dim_n)
```

See https://output.circle-artifacts.com/output/job/3b105100-89ce-43d8-8cbb-b3c68f3daac7/artifacts/0/html/reference/generated/scipy.integrate.cubature.html#scipy.integrate.cubature